src/cluster_migrate.c

```diff
@@ -0,0 +1,1950 @@
+ *            ┌─────────────────────────────┐      │
+ *            │SLOT_MIGRATION_JOB_CANCELLED│◄─────┘
+ *            └─────────────────────────────┘
+ *
+ */
+
+/* Returns 1 if the given slot is being exported, 0 otherwise. */
+int clusterIsSlotExporting(int slot) {
+    listNode *ln;
+    listIter li;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationJob *job = ln->value;
+        if (job->type != SLOT_MIGRATION_EXPORT) continue;
+        if (isSlotMigrationJobFinished(job)) continue;
+        if (isSlotInSlotRanges(slot, job->slot_ranges)) return 1;
+    }
+    return 0;
+}
+
+/* Returns the name of the slot export target for the given slot, or NULL if the
+ * slot is not being exported. */
+char *getNameOfSlotExportTarget(int slot) {
+    listNode *ln;
+    listIter li;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationJob *job = ln->value;
+        if (job->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!isSlotMigrationJobInProgress(job)) continue;
+        if (isSlotInSlotRanges(slot, job->slot_ranges)) return job->nodename;
+    }
+    return NULL;
+}
+
+/* Sent by an operator to the current owner of one or more slot ranges. The
+ * source will attempt to migrate the slot ranges to the specified target node. */
+void clusterCommandMigrate(client *c) {
+    if (!nodeIsPrimary(server.cluster->myself)) {
+        addReplyError(c, "Slot migration can only be used on primary nodes.");
+        return;
+    }
+
+    if (isAnySlotInManualImportingState()) {
+        addReplyError(c, "Some slots are being manually imported. "
+                         "Please get all slots to a stable state before attempting import.");
+        return;
+    }
+    if (isAnySlotInManualMigratingState()) {
+        addReplyError(c, "Some slots are being manually migrated. "
+                         "Please get all slots to a stable state before attempting import.");
+        return;
+    }
+
+    int curr_index = 2;
+    list *new_slot_migrations = listCreate();
+    listSetFreeMethod(new_slot_migrations, freeSlotMigrationJob);
+
+    while (curr_index < c->argc) {
+        if (strcasecmp(c->argv[curr_index]->ptr, "slotsrange")) {
+            addReplyErrorObject(c, shared.syntaxerr);
+            listRelease(new_slot_migrations);
+            return;
+        }
+        curr_index++;
+
+        clusterNode *source_node = NULL;
+        list *slot_ranges = parseSlotRangesOrReply(c, curr_index, &curr_index, &source_node);
+        if (slot_ranges == NULL) {
+            listRelease(new_slot_migrations);
+            return;
+        }
+        if (source_node != server.cluster->myself) {
+            addReplyErrorFormat(c, "Slots are not served by myself.");
+            listRelease(slot_ranges);
+            listRelease(new_slot_migrations);
+            return;
```

Comment:
similar comments - can we break out the loop and free these resources just once before returning?